scripts/find-or-make-bucket.sh

```diff
@@ -0,0 +1,94 @@
+
+source ./scripts/common.sh
+
+# This script examines the current environment and determines whether to check for an
+# existing, previously tested bucket or to make a new one. It expects to be run on GitHub
+# Actions push events only.
+
+build_and_sync_bucket() {
+    echo "Making a new bucket..."
+    ./scripts/build-site.sh
+    ./scripts/bucketize.sh
+}
+
+get_bucket_for_commit() {
+    aws ssm get-parameter \
+        --name "$(ssm_parameter_key_for_commit $1)" \
+        --query Parameter.Value \
+        --region us-west-2 \
+        --output text || echo ""
+}
+
+get_associated_pr_commit_for_commit() {
+    # Note that this uses a GitHub preview API.
+    # https://docs.github.com/en/rest/reference/repos#list-pull-requests-associated-with-a-commit
+    curl -H "Accept: application/vnd.github.groot-preview+json" \
+        "https://api.github.com/repos/${GITHUB_REPOSITORY}/commits/$1/pulls" | jq -r '.[0].head.sha' || echo ""
+}
+
+# For push events, the GITHUB_SHA is the HEAD of the base branch (e.g., master).
+# https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables
+if [[ "$GITHUB_EVENT_NAME" == "push" && ! -z "$GITHUB_REPOSITORY" &&  ! -z "$GITHUB_SHA" ]]; then
+
+    # Attempt to find the most recent PR commit associated with the commit that triggered this workflow.
+    associated_pr_commit="$(get_associated_pr_commit_for_commit $GITHUB_SHA)"
+
+    # If there's an associated PR commit, attempt to find the bucket that goes with it.
+    if [ "$associated_pr_commit" != "null" ]; then
+
+        # All commits originating from PRs *should* have an associated bucket, but some
+        # may not -- for example, if a merge occurred before the PR build completed, or if
+        # the commit is was the result of a direct push to master. In these cases, we
+        # build and test a new bucket before deploying it in a later step. If we do find a
+        # bucket associated with the commit, however, we run a few checks, and if things
+        # look good, we exit, leaving its metadata file in place for the Pulumi update.
+
+        # Query AWS Parameter Store for a bucket associated with the referenced commit.
+        candidate_bucket="$(get_bucket_for_commit $associated_pr_commit)"
+
+        # If a matching bucket wasn't found, just make one. Otherwise, continue.
+        if [ -z "$candidate_bucket" ]; then
+            echo "Push event detected, but there doesn't seem to be a bucket associated with the referenced commit ${associated_pr_commit}."
+            build_and_sync_bucket
+            exit 0
+        fi
+
+        # Validate that the bucket exists and has a metadata file, then download that file
+        # to the local path that Pulumi will be expecting. We'll use this file to run the
+        # Pulumi update if it passes all of our checks.
+        aws s3 cp s3://${candidate_bucket}/metadata.json $(origin_bucket_metadata_filepath)
```

Comment:
What do you think about adding this file as a GH Action [artifact](https://docs.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts)? It could be useful to look at this file when looking at a specific workflow execution to see what it produced. You could download this file as part of a `push` build and avoid the SSM param thing altogether most likely.

Reply:
Definitely like the idea of producing it as a built artifact, yeah, for reference. I'll look into adding that. 

I'd still need the ability to look up the bucket associated with a particular commit later, so having it stored solely as an artifact would make that a good deal more cumbersome, and since I want to run the whole job to succeed or fail in one step, rather than break it out into separate steps, there wouldn't be much benefit to producing and then consuming the artifact as part of the workflow itself. So as a replacement for SSM for PRs and pushes, it's not quite the right fit. Less frequently, there'll also be the occasionaly need to produce these files and use them for deployments outside the context of an Actions workflow (e.g., while developing a dev stack, deploying to prod in an emergency, etc.).

But yeah, I do like having it for reference -- it'll most likely make deleting the bucket associated with a closed-unmerged PR (which I'm working on now) way easier. 🤞 So thanks for the tip!

Reply:
Added the metadata file and Cypress recordings as artifacts. Thanks again!